mypy/checkexpr.py

```diff
@@ -1840,89 +1828,234 @@ def check_op_local(self, method: str, base_type: Type, arg: Expression,
+                and isinstance(right_type, Instance)
+                and lookup_definer(left_type, op_name) != lookup_definer(right_type, rev_op_name)):
+            # When we do "A() + B()" where B is a subclass of B, we'll actually try calling
+            # B's __radd__ method first, but ONLY if B explicitly defines or overrides the
+            # __radd__ method.
+            #
+            # This mechanism lets subclasses "refine" the expected outcome of the operation, even
+            # if they're located on the RHS.
+
+            variants_raw.append((right_op, right_type, left_expr))
+            variants_raw.append((left_op, left_type, right_expr))
+        else:
+            # In all other cases, we do the usual thing and call __add__ first and
+            # __radd__ second when doing "A() + B()".
+
+            variants_raw.append((left_op, left_type, right_expr))
+            variants_raw.append((right_op, right_type, left_expr))
+
+        # STEP 2b:
+        # When running Python 2, we might also try calling the __cmp__ method.
+
+        is_python_2 = self.chk.options.python_version[0] == 2
+        if is_python_2 and op_name in nodes.ops_falling_back_to_cmp:
+            cmp_method = nodes.comparison_fallback_method
+            left_cmp_op = lookup_operator(cmp_method, left_type)
+            right_cmp_op = lookup_operator(cmp_method, right_type)
+
+            if bias_right:
+                variants_raw.append((right_cmp_op, right_type, left_expr))
+                variants_raw.append((left_cmp_op, left_type, right_expr))
+            else:
+                variants_raw.append((left_cmp_op, left_type, right_expr))
+                variants_raw.append((right_cmp_op, right_type, left_expr))
+
+        # STEP 3:
+        # We now filter out all non-existant operators. The 'variants' list contains
+        # all operator methods that are actually present, in the order that Python
+        # attempts to invoke them.
+
+        variants = [(op, obj, arg) for (op, obj, arg) in variants_raw if op is not None]
+
+        # STEP 4:
+        # We now try invoking each one. If an operation succeeds, end early and return
+        # the corresponding result. Otherwise, return the result and errors associated
+        # with the first entry.
+
+        errors = []
+        results = []
+        for method, obj, arg in variants:
+            local_errors = make_local_errors()
+
+            callable_name = None  # type: Optional[str]
+            if isinstance(obj, Instance):
+                # TODO: Find out in which class the method was defined originally?
+                # TODO: Support non-Instance types.
+                callable_name = '{}.{}'.format(obj.type.fullname(), op_name)
+
+            result = self.check_call(method, [arg], [nodes.ARG_POS],
+                                   context, arg_messages=local_errors,
+                                   callable_name=callable_name, object_type=obj)
```

Comment:
Can this be refactored to use `check_op_local `, it looks like there is some code (and TODO comments) duplication.